File: python/day03.py
from aocd import get_data
from dotenv import load_dotenv

load_dotenv()


def find_common_item_one(contents: str) -> str:
    split_point = len(contents) // 2
    compartment1, compartment2 = contents[:split_point], contents[split_point:]
    return next(iter(set(compartment1).intersection(set(compartment2))))


def find_common_item_three(contents: list[str]) -> str:
    print(contents)
    return next(iter(set(contents[0]).intersection(set(contents[1]), set(contents[2]))))


def priority(char: str) -> int:
    num = ord(char)
    return num - 96 if num > 90 else num - 38


def day03_a(all_contents: str) -> int:
    return sum(map(lambda x: priority(find_common_item_one(x)), all_contents.split("\n")))


def day03_b(all_contents: str) -> int:
    rucksack_contents = all_contents.split("\n")
    return sum(map(lambda x: priority(find_common_item_three(rucksack_contents[x:x + 3])),
                   range(0, len(rucksack_contents), 3)))


if __name__ == '__main__':
    data = get_data(session, day=3, year=2022)
    print(day03_a(data))
    print(day03_b(data))

File: rust/day03.rs
use std::fs;
use std::collections::HashSet;
use std::iter::FromIterator;


fn get_priority(c: &char) -> u32 {
    let v: u32 = *c as u32;
    return if v > 90 {
        v - 96
    } else {
        v - 38
    }
}

fn day03_a(data: &str) -> u32 {
    let mut priority_sum: u32 = 0;

    for rucksack in data.split("\n") {
        let (comp1, comp2): (&str, &str) = rucksack.split_at(rucksack.len() / 2);
        let (comp1set, comp2set): (HashSet<char>, HashSet<char>) = (HashSet::from_iter(comp1.chars()), HashSet::from_iter(comp2.chars()));
        let intersection: Vec<_> = comp1set.intersection(&comp2set).collect();
        priority_sum += get_priority(intersection[0]);
    }

    return priority_sum;

}


fn day03_b(data: &str) -> u32 {
    let mut priority_sum: u32 = 0;
    let rucksacks: Vec<_> = data.split("\n").collect();
    for i in (0..(rucksacks.len()- 3)).step_by(3) {
        let (comp1, comp2, comp3): (&str, &str, &str) = (rucksacks[i], rucksacks[i+1], rucksacks[i+2]);
        let (comp1set, comp2set, comp3set): (HashSet<char>, HashSet<char>,  HashSet<char>) =
            (HashSet::from_iter(comp1.chars()), HashSet::from_iter(comp2.chars()), HashSet::from_iter(comp3.chars()));
        let sets = [comp1set, comp2set, comp3set];
        let intersection = sets
        .iter()
        .skip(1)
        .fold(sets[0].clone(), |acc, hs| {
            acc.intersection(hs).cloned().collect()
        });
        priority_sum += get_priority(&intersection.iter().next().unwrap().clone());

    }
    return priority_sum;
}


fn main() {
    let contents = fs::read_to_string("inputs/day03.txt")
        .expect("Should have been able to read the file");
    let priorities_a = day03_a(&contents);
    println!("Sum of priorities (part a): {}", priorities_a);
    let priorities_b = day03_b(&contents);
    println!("Sum of priorities (part b): {}", priorities_b);
}